lib/pure/asyncmacro.nim

```diff
@@ -221,13 +221,11 @@ proc asyncSingleProc(prc: NimNode): NimNode =
     procBody = newStmtList()
     let resultIdent = ident"result"
     procBody.add quote do:
-      template nimAsyncDispatchSetResult(x: `subRetType`) {.used.} =
-        # If the proc has implicit return then this will get called
-        `resultIdent` = x
-      template nimAsyncDispatchSetResult(x: untyped) {.used.} =
-        # If the proc doesn't have implicit return then this will get called
-        x
-    procBody.add newCall(ident"nimAsyncDispatchSetResult", blockStmt)
+      # Check whether there is an implicit return
+      when typeof(`blockStmt`) is void:
```

Comment:
does it matter that the AST essentially has 3 references to the same AST here? ie the cgen has an ugly habit to modify the AST for example which potentially could be brittle.

Reply:
Shouldn't the `when` cause only one reference to get to the cgen stage?

Reply:
The cgen doesn't modify the AST. Well it does and we need to get rid of that hack, but it doesn't modify the AST in a way that hurts here.